File: cs/src/DataCentric/Types/Record/Key.cs
/*
Copyright (C) 2013-present The DataCentric Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Text;
using MongoDB.Bson;
using NodaTime;

namespace DataCentric
{
    /// <summary>
    /// Base class of a foreign key.
    ///
    /// The curiously recurring template pattern (CRTP) key class
    /// Key(TKey,TRecord) is derived from this class.
    ///
    /// Any elements of defined in the type specific key record
    /// become key tokens. Property Value and method ToString() of
    /// the key consists of key tokens with semicolon delimiter.
    /// </summary>
    public abstract class Key : Data
    {
        /// <summary>
        /// String key consists of semicolon delimited primary key elements:
        ///
        /// KeyElement1;KeyElement2
        ///
        /// To avoid serialization format uncertainty, key elements
        /// can have any atomic type except Double.
        /// </summary>
        public string Value
        {
            get
            {
                var tokens = new List<string>();
                var elementInfoArray = DataTypeInfo.GetOrCreate(GetType()).DataElements;
                foreach (var elementInfo in elementInfoArray)
                {
                    // Convert key element to string key token.
                    //
                    // Note that string representation of certain types inside the key
                    // is not the same as what is returned by AsString().
                    //
                    // Specifically, LocalDate and LocalTime are represented as tokens
                    // in readable int format without delimiters (yyyymmdd and hhmmssfff)
                    // but use delimited ISO format (yyyy-mm-dd and hh:mm:ss.fff) when
                    // serialized using AsString().
                    var token = GetKeyToken(this, elementInfo);
                    tokens.Add(token);
                }

                string result = string.Join(";", tokens);
                return result;
            }
        }

        /// <summary>
        /// Convert key element to string key token.
        ///
        /// Note that string representation of certain types inside the key
        /// is not the same as what is returned by AsString().
        ///
        /// Specifically, LocalDate and LocalTime are represented as tokens
        /// in readable int format without delimiters (yyyymmdd and hhmmssfff)
        /// but use delimited ISO format (yyyy-mm-dd and hh:mm:ss.fff) when
        /// serialized using AsString().
        /// </summary>
        internal static string GetKeyToken(object obj, PropertyInfo elementInfo)
        {
            string result;
            object element = elementInfo.GetValue(obj);
            switch (element)
            {
                case null:
                    throw new Exception($"Key element {elementInfo.Name} of type {obj.GetType().Name} is null. " +
                                        $"Null elements are not permitted in key.");
                case string stringValue:
                    // Return the string after checking that it is not empty
                    // and that it does not itself contain semicolon delimiters
                    if (stringValue == String.Empty) throw new Exception(
                            $"String key element {elementInfo.Name} is empty." +
                            $"Empty elements are not permitted in key.");
                    if (stringValue.Contains(";")) throw new Exception(
                        $"Key element {elementInfo.Name} of type {obj.GetType().Name} includes semicolon delimiter. " +
                        $"The use of this delimiter is reserved for separating key tokens.");
                    result = stringValue;
                    break;
                case double doubleValue:
                    // Double key elements are not permitted due to the risk of
                    // key mismatch as a result of different serialization settings
                    // across multiple libraries and programming languages
                    throw new Exception(
                        $"Key element {elementInfo.Name} of type {obj.GetType().Name} has type Double. Elements " +
                        $"of this type cannot be part of key due to serialization format uncertainty.");
                case LocalDate dateValue:
                    // Serialize as readable int in yyyymmdd format, not in delimited ISO format
                    result = dateValue.ToIsoInt().ToString();
                    break;
                case LocalTime timeValue:
                    // Serialize as readable int in hhmmssfff format, not in delimited ISO format
                    result = timeValue.ToIsoInt().ToString();
                    break;
                case LocalMinute minuteValue:
                    // Serialize as readable int in hhmm format, not in delimited ISO format
                    result = minuteValue.ToIsoInt().ToString();
                    break;
                case LocalDateTime dateTimeValue:
                    // Serialize as readable long in yyyymmddhhmmssfff format, not in delimited ISO format
                    result = dateTimeValue.ToIsoLong().ToString();
                    break;
                case bool boolValue:
                case int intValue:
                case long longValue:
                case RecordId objectIdValue:
                case Enum enumValue:
                case Key keyValue:
                    // Use AsString() for all remaining types including the key
                    //
                    // A token representing another key can contain a semicolon delimiter
                    // if the key being converted to the token is composite (has more than
                    // one key element). However a string key element cannot contain the
                    // semicolon delimiter.
                    result = element.AsString();
                    break;
                default:
                    // Argument type is unsupported, error message
                    throw new Exception(
                        $"Key element {elementInfo.Name} of type {obj.GetType().Name} has type {element.GetType()}" +
                        $"that is not one of the supported key element types. Available key element types are " +
                        $"string, double, bool, int, long, LocalDate, LocalTime, LocalMinute, LocalDateTime, LocalMinute, RecordId, or Enum.");
            }

            return result;
        }

        /// <summary>
        /// String key consists of semicolon delimited primary key elements:
        ///
        /// KeyElement1;KeyElement2
        ///
        /// To avoid serialization format uncertainty, key elements
        /// can have any atomic type except Double.
        /// </summary>
        public override string ToString() { return Value; }

        /// <summary>
        /// Populate key elements from semicolon delimited string.
        /// Elements that are themselves keys may use more than
        /// one token.
        ///
        /// If key AKey has two elements, B and C, where
        ///
        /// * B has type BKey which has two string elements, and
        /// * C has type string,
        ///
        /// the semicolon delimited key has the following format:
        ///
        /// BToken1;BToken2;CToken
        ///
        /// To avoid serialization format uncertainty, key elements
        /// can have any atomic type except Double.
        /// </summary>
        public void PopulateFrom(string value)
        {
            // Split key into tokens
            var tokens = value.Split(';');

            // Call the private method that uses array of tokens.
            // This method returns the number of tokens actually
            // used to parse the key.
            int tokenIndex = PopulateFrom(tokens, 0);

            // Verify that all tokens have been used, error message otherwise
            if (tokens.Length != tokenIndex)
            {
                throw new Exception($"Key with type {GetType().Name} requires {tokenIndex} tokens including " +
                                    $"any composite key elements, while key value {value} contains {tokens.Length} tokens.");
            }
        }

        //--- PRIVATE

        /// <summary>
        /// Populate key elements from an array of tokens starting
        /// at the specified token index. Elements that are themselves
        /// keys may use more than one token.
        ///
        /// This method returns the index of the first unused token.
        /// The returned value is the same as the length of the tokens
        /// array if all tokens are used.
        ///
        /// If key AKey has two elements, B and C, where
        ///
        /// * B has type BKey which has two string elements, and
        /// * C has type string,
        ///
        /// the semicolon delimited key has the following format:
        ///
        /// BToken1;BToken2;CToken
        ///
        /// To avoid serialization format uncertainty, key elements
        /// can have any atomic type except Double.
        /// </summary>
        private int PopulateFrom(string[] tokens, int tokenIndex)
        {
            // Get key elements using reflection
            var elementInfoArray = DataTypeInfo.GetOrCreate(this).DataElements;

            // If singleton is detected process it separately, then exit
            if (elementInfoArray.Length == 0)
            {
                // Check that string key is empty
                if (tokens.Length != 1 || tokens[0] != String.Empty)
                    throw new Exception($"Type {GetType()} has key {string.Join(";", tokens)} while " +
                                        $"for a singleton the key must be an empty string (String.Empty). " +
                                        $"Singleton key is a key that has no key elements.");

                // Return the length of empty key which consists of one (empty) token
                return 1;
            }

            // Check that there are enough remaining tokens in the key for each key element
            if (tokens.Length - tokenIndex < elementInfoArray.Length)
            {
                throw new Exception(
                    $"Key of type {GetType().Name} requires at least {elementInfoArray.Length} elements " +
                    $"{String.Join(";", elementInfoArray.Select(p => p.Name).ToArray())} while there are " +
                    $"only {tokens.Length - tokenIndex} remaining key tokens: {string.Join(";", tokens)}.");
            }

            // Iterate over element info elements, advancing tokenIndex by the required
            // number of tokens for each element. In case of embedded keys, the value of
            // tokenIndex is advanced by the recursive call to InitFromTokens method
            // of the embedded key.
            foreach (var elementInfo in elementInfoArray)
            {
                // Get element type
                Type elementType = elementInfo.PropertyType;

                // Convert string token to value depending on elementType
                if (elementType == typeof(string))
                {
                    CheckTokenNotEmpty(tokens, tokenIndex);

                    string token = tokens[tokenIndex++];
                    elementInfo.SetValue(this, token);
                }
                else if (elementType == typeof(double) || elementType == typeof(double?))
                {
                    throw new Exception(
                        $"Key element {elementInfo.Name} has type Double. Elements of this type " +
                        $"cannot be part of key due to serialization format uncertainty.");
                }
                else if (elementType == typeof(bool) || elementType == typeof(bool?))
                {
                    CheckTokenNotEmpty(tokens, tokenIndex);

                    string token = tokens[tokenIndex++];
                    bool tokenValue = bool.Parse(token);
                    elementInfo.SetValue(this, tokenValue);
                }
                else if (elementType == typeof(int) || elementType == typeof(int?))
                {
                    CheckTokenNotEmpty(tokens, tokenIndex);

                    string token = tokens[tokenIndex++];
                    int tokenValue = int.Parse(token);
                    elementInfo.SetValue(this, tokenValue);
                }
                else if (elementType == typeof(long) || elementType == typeof(long?))
                {
                    CheckTokenNotEmpty(tokens, tokenIndex);

                    string token = tokens[tokenIndex++];
                    long tokenValue = long.Parse(token);
                    elementInfo.SetValue(this, tokenValue);
                }
                else if (elementType == typeof(LocalDate) || elementType == typeof(LocalDate?))
                {
                    CheckTokenNotEmpty(tokens, tokenIndex);

                    // Inside the key, LocalDate is represented as readable int in
                    // non-delimited yyyymmdd format, not as delimited ISO string.
                    //
                    // First parse the string to int, then convert int to LocalDate.
                    string token = tokens[tokenIndex++];
                    if (!Int32.TryParse(token, out int isoInt))
                    {
                        throw new Exception(
                            $"Element {elementInfo.Name} of key type {GetType().Name} has type LocalDate and value {token} " +
                            $"that cannot be converted to readable int in non-delimited yyyymmdd format.");
                    }

                    LocalDate tokenValue = LocalDateImpl.ParseIsoInt(isoInt);
                    elementInfo.SetValue(this, tokenValue);
                }
                else if (elementType == typeof(LocalTime) || elementType == typeof(LocalTime?))
                {
                    CheckTokenNotEmpty(tokens, tokenIndex);

                    // Inside the key, LocalTime is represented as readable int in
                    // non-delimited hhmmssfff format, not as delimited ISO string.
                    //
                    // First parse the string to int, then convert int to LocalTime.
                    string token = tokens[tokenIndex++];
                    if (!Int32.TryParse(token, out int isoInt))
                    {
                        throw new Exception(
                            $"Element {elementInfo.Name} of key type {GetType().Name} has type LocalTime and value {token} " +
                            $"that cannot be converted to readable int in non-delimited hhmmssfff format.");
                    }

                    LocalTime tokenValue = LocalTimeImpl.ParseIsoInt(isoInt);
                    elementInfo.SetValue(this, tokenValue);
                }
                else if (elementType == typeof(LocalMinute) || elementType == typeof(LocalMinute?))
                {
                    CheckTokenNotEmpty(tokens, tokenIndex);

                    // Inside the key, LocalMinute is represented as readable int in
                    // non-delimited hhmm format, not as delimited ISO string.
                    //
                    // First parse the string to int, then convert int to LocalTime.
                    string token = tokens[tokenIndex++];
                    if (!Int32.TryParse(token, out int isoInt))
                    {
                        throw new Exception(
                            $"Element {elementInfo.Name} of key type {GetType().Name} has type LocalMinute and value {token} " +
                            $"that cannot be converted to readable int in non-delimited hhmm format.");
                    }

                    LocalMinute tokenValue = LocalMinuteImpl.ParseIsoInt(isoInt);
                    elementInfo.SetValue(this, tokenValue);
                }
                else if (elementType == typeof(LocalDateTime) || elementType == typeof(LocalDateTime?))
                {
                    CheckTokenNotEmpty(tokens, tokenIndex);

                    // Inside the key, LocalDateTime is represented as readable long in
                    // non-delimited yyyymmddhhmmssfff format, not as delimited ISO string.
                    //
                    // First parse the string to long, then convert int to LocalDateTime.
                    string token = tokens[tokenIndex++];
                    if (!Int64.TryParse(token, out long isoLong))
                    {
                        throw new Exception(
                            $"Element {elementInfo.Name} of key type {GetType().Name} has type LocalDateTime and value {token} " +
                            $"that cannot be converted to readable long in non-delimited yyyymmddhhmmssfff format.");
                    }

                    LocalDateTime tokenValue = LocalDateTimeImpl.ParseIsoLong(isoLong);
                    elementInfo.SetValue(this, tokenValue);
                }
                else if (elementType == typeof(RecordId) || elementType == typeof(RecordId?))
                {
                    CheckTokenNotEmpty(tokens, tokenIndex);

                    string token = tokens[tokenIndex++];
                    RecordId tokenValue = RecordId.Parse(token);
                    elementInfo.SetValue(this, tokenValue);
                }
                else if (elementType.BaseType == typeof(Enum)) // TODO Support nullable Enum in key
                {
                    CheckTokenNotEmpty(tokens, tokenIndex);

                    string token = tokens[tokenIndex++];
                    object tokenValue = Enum.Parse(elementType, token);
                    elementInfo.SetValue(this, tokenValue);
                }
                else if (typeof(Key).IsAssignableFrom(elementType))
                {
                    Key keyElement = (Key)Activator.CreateInstance(elementType);
                    tokenIndex = keyElement.PopulateFrom(tokens, tokenIndex);
                    elementInfo.SetValue(this, keyElement);
                }
                else
                {
                    // Field type is unsupported for a key, error message
                    throw new Exception(
                        $"Element {elementInfo.Name} of key type {GetType().Name} has type {elementType} that " +
                        $"is not one of the supported key element types. Available key element types are " +
                        $"string, bool, int, long, LocalDate, LocalTime, LocalMinute, LocalDateTime, Enum, or Key.");
                }
            }

            return tokenIndex;
        }

        /// <summary>
        /// Check that token at the specified position is not empty.
        /// </summary>
        private void CheckTokenNotEmpty(string[] tokens, int tokenIndex)
        {
            string token = tokens[tokenIndex];
            if (string.IsNullOrEmpty(token))
                throw new Exception($"Key {string.Join(";", tokens)} for key type {GetType().Name} contains an empty token.");
        }
    }

    /// <summary>Extension methods for Key.</summary>
    public static class KeyExt
    {
        /// <summary>
        /// Deserialize record from XML using short class name without namespace
        /// for the root XML element.
        /// </summary>
        public static void ParseXml(this Key obj, string xmlString)
        {
            ITreeReader reader = new XmlTreeReader(xmlString);

            // Root node of serialized XML must be the same as mapped class name without namespace
            var mappedFullName = ClassInfo.GetOrCreate(obj).MappedClassName;
            ITreeReader recordNodes = reader.ReadElement(mappedFullName);

            // Deserialize from XML nodes inside the root node
            obj.DeserializeFrom(recordNodes);
        }

        /// <summary>
        /// Serialize record to XML using short class name without namespace
        /// for the root XML element.
        /// </summary>
        public static string ToXml(this Key obj)
        {
            // Get root XML element name using mapped final type of the object
            string rootName = ClassInfo.GetOrCreate(obj).MappedClassName;

            // Serialize to XML
            ITreeWriter writer = new XmlTreeWriter();
            writer.WriteStartDocument(rootName);
            obj.SerializeTo(writer);
            writer.WriteEndDocument(rootName);

            // Convert to string
            string result = writer.ToString();
            return result;
        }
    }
}

File: py/src/datacentric/types/record/key.py
import datetime as dt
from abc import ABC
from enum import Enum
from typing import List
from bson import ObjectId

from datacentric.types.time.local_minute import LocalMinute
from datacentric.types.record import Data
import datacentric.types.time.date_ext as date_ext


class Key(Data, ABC):
    """Base class of a foreign key. Any slots defined in
    type specific become key tokens. Property value and str(self)
    consists of key tokens with semicolon delimiter.
    """
    __slots__ = []

    def __init__(self):
        super().__init__()

    def __str__(self):
        return self.value

    @property
    def value(self) -> str:
        """String key consists of semicolon delimited primary key elements:

        key_element1;key_element2

        To avoid serialization format uncertainty, key elements
        can have any atomic type except float.
        """
        tokens = []
        element_array = self.__slots__
        if type(self.__slots__) is str:
            element_array = [element_array]

        for element in element_array:
            tokens.append(self.get_key_token(self, element))
        return ';'.join(tokens)

    @staticmethod
    def get_key_token(obj: object, slot: str) -> str:
        """Convert key element to string key token.
        """
        attr_value = obj.__getattribute__(slot)
        attr_type = type(attr_value)
        if attr_value is None:
            raise ValueError(f'Key element {slot} of type {type(obj).__name__} is null. '
                             f'Null elements are not permitted in key.')
        elif attr_type == str:
            if attr_value == '':
                raise ValueError(f'String key element {slot} is empty. Empty elements are not permitted in key.')
            if ';' in attr_value:
                raise ValueError(f'Key element {slot} of type {type(obj).__name__} includes semicolon delimiter. '
                                 f'The use of this delimiter is reserved for separating key tokens.')
            return attr_value
        elif attr_type == float:
            raise ValueError(f'Key element {slot} of type {type(obj).__name__} has type float. '
                             f'Elements of this type cannot be part of key due to serialization format uncertainty.')
        elif attr_type == dt.date:
            return str(date_ext.date_to_iso_int(attr_value))
        elif attr_type == dt.time:
            return str(date_ext.time_to_iso_int(attr_value))
        elif attr_type == LocalMinute:
            return str(date_ext.minute_to_iso_int(attr_value))
        elif attr_type == dt.datetime:
            return str(date_ext.date_time_to_iso_int(attr_value))
        elif attr_type == bool:
            if attr_value:
                return 'true'
            else:
                return 'false'
        elif attr_type == int:
            return str(attr_value)
        elif attr_type == ObjectId:
            return str(attr_value)
        elif issubclass(attr_type, Key):
            return str(attr_value)
        elif issubclass(attr_type, Enum):
            return attr_value.name
        else:
            raise ValueError(f'Key element {slot} of type {type(obj).__name__} has type {attr_type.__name__} '
                             f'that is not one of the supported key element types. Available key element types are '
                             f'string, double, bool, int, long, LocalDate, LocalTime, LocalMinute, LocalDateTime, '
                             f'LocalMinute, ObjectId, or Enum.')

    def populate_from_string(self, value: str) -> None:
        """Populate key attributes from semicolon delimited string.
        Attributes that are themselves keys may use more than
        one token.

        If key AKey has two elements, B and C, where

        * B has type BKey which has two string elements, and
        * C has type string,

        the semicolon delimited key has the following format:

        BToken1;BToken2;CToken
        """
        tokens = value.split(';')
        token_index = self.__populate_from_string(tokens, 0)

        if len(tokens) != token_index:
            raise Exception(f'Key with type {type(self).__name__} requires {token_index} tokens including '
                            f'any composite key elements, while key value {value} contains {len(tokens)} tokens.')

    def __populate_from_string(self, tokens: List[str], token_index: int) -> int:
        """Populate key elements from an array of tokens starting
        at the specified token index.
        """
        slots = self.__slots__
        if type(slots) is str:
            slots = [slots]

        # Singleton key case
        if len(slots) == 0:
            if len(tokens) != 1 or tokens[0] != '':
                raise Exception(f'Type {type(self).__name__} has key {";".join(tokens)} while '
                                f'for a singleton the key must be an empty string. '
                                f'Singleton key is a key that has no key elements.')
            return 1
        if len(tokens) - token_index < len(slots):
            raise Exception(f'Key of type {type(self).__name__} requires at least {len(slots)} elements '
                            f'{";".join(slots)} while there are only {len(tokens) - token_index} remaining key tokens:'
                            f'{";".join(tokens)}')

        """
        Iterate over element info elements, advancing tokenIndex by the required
        number of tokens for each element. In case of embedded keys, the value of
        tokenIndex is advanced by the recursive call to InitFromTokens method
        of the embedded key.
        """
        annotations = type(self).__annotations__
        for slot in slots:
            member_type = annotations[slot]
            if member_type == type(float):
                raise Exception(f'Key element {slot} has type Double. Elements of this type '
                                f'cannot be part of key due to serialization format uncertainty.')
            if issubclass(member_type, Key):
                key_element = member_type()
                token_index = key_element.__populate_from_string(tokens, token_index)
                self.__setattr__(slot, key_element)
                continue

            # Check that token is not empty
            token = tokens[token_index]
            if token == '':
                raise Exception(f'Key {";".join(tokens)} for key type {type(self).__name__} contains an empty token.')

            if member_type == str:
                value = token
            elif member_type == bool:
                value = bool(token.capitalize())
            elif member_type == int:
                value = int(token)
            elif member_type == dt.date:
                value = date_ext.iso_int_to_date(int(token))
            elif member_type == dt.time:
                value = date_ext.iso_int_to_time(int(token))
            elif member_type == LocalMinute:
                value = date_ext.iso_int_to_local_minute(int(token))
            elif member_type == dt.datetime:
                value = date_ext.iso_int_to_date_time(int(token))
            elif member_type == ObjectId:
                value = ObjectId(token)
            elif issubclass(member_type, Enum):
                value = member_type[token]
            else:
                raise Exception(f'Unexpected type {member_type.__name__} in key tokens.')

            self.__setattr__(slot, value)
            token_index += 1

        return token_index
